controller/dataplane/konnect_extension.go

```diff
@@ -0,0 +1,104 @@
+package dataplane
+
+import (
+	"context"
+	"errors"
+
+	"github.com/samber/lo"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	operatorv1alpha1 "github.com/kong/gateway-operator/api/v1alpha1"
+	"github.com/kong/gateway-operator/api/v1beta1"
+	dputils "github.com/kong/gateway-operator/internal/utils/dataplane"
+	"github.com/kong/gateway-operator/pkg/consts"
+	k8sutils "github.com/kong/gateway-operator/pkg/utils/kubernetes"
+	k8sresources "github.com/kong/gateway-operator/pkg/utils/kubernetes/resources"
+)
+
+// applyDataPlaneKonnectExtension gets the DataPlane as argument, and in case it references a KonnectExtension, it
+// fetches the referenced extension and applies the necessary changes to the DataPlane spec.
+func applyDataPlaneKonnectExtension(ctx context.Context, cl client.Client, dataplane *v1beta1.DataPlane) error {
+	for _, extensionRef := range dataplane.Spec.Extensions {
+		if extensionRef.Group != operatorv1alpha1.SchemeGroupVersion.Group || extensionRef.Kind != operatorv1alpha1.DataPlaneKonnectExtensionKind {
+			continue
+		}
+		namespace := dataplane.Namespace
+		if extensionRef.Namespace != nil && *extensionRef.Namespace != namespace {
+			return errors.New("cross-namespace reference is not currently supported for Konnect extensions")
+		}
+
+		konnectExt := operatorv1alpha1.DataPlaneKonnectExtension{}
+		if err := cl.Get(ctx, client.ObjectKey{
+			Namespace: namespace,
+			Name:      extensionRef.Name,
+		}, &konnectExt); err != nil {
+			return err
+		}
+
+		if dataplane.Spec.Deployment.PodTemplateSpec == nil {
+			dataplane.Spec.Deployment.PodTemplateSpec = &corev1.PodTemplateSpec{}
+		}
+
+		d := k8sresources.Deployment(appsv1.Deployment{
+			Spec: appsv1.DeploymentSpec{
+				Template: *dataplane.Spec.Deployment.PodTemplateSpec,
+			},
+		})
+		if container := k8sutils.GetPodContainerByName(&d.Spec.Template.Spec, consts.DataPlaneProxyContainerName); container == nil {
+			d.Spec.Template.Spec.Containers = append(d.Spec.Template.Spec.Containers, corev1.Container{
+				Name: consts.DataPlaneProxyContainerName,
+			})
+		}
+
+		d.WithVolume(kongInKonnectClusterCertificateVolume())
+		d.WithVolumeMount(kongInKonnectClusterCertificateVolumeMount(), consts.DataPlaneProxyContainerName)
+		d.WithVolume(kongInKonnectClusterCertVolume(konnectExt.Spec.AuthConfiguration.ClusterCertificateSecretRef.Name))
```

Comment:
Is the existence of the referred secret checked at any point and communicated to the user in e.g. `DataPlane`'s status conditions that it's e.g. missing?

Reply:
That part was not implemented, but it's actually a good point. I implemented the `ResolvedRefs` condition for the `DataPlane` in e6ba10117baf8f3c43f9b50a0db6f867f9bbda2c